README.md
# privteGPTServer
a fast-api server to talk with state-of-the-art Open source LLM Models with/out your custom data with just **2-3 GB** RAM.


## Introductions
privategptserver is using state-of-the-art open source large language models (as of now it's `falcon7B`) allowing users   
to opt in run the server and call the LLM with api calls or integrate it with your own data   
by simply upload your pdf file and chat with it **without the need for powerfull GPUs or CPUs**

> Note: this not near a final product,this still work in progress and it's in the very early stage but it works, I plan to first  
>  easing the user experance to run the server then optimize it, adding more LLM  
>  and would be more than happy to accpet any contribution  
## Demo
this a demo of what to expect using this server  [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1pRNiVFJs5uJ5OZnhqpJ4sauTL53wtR1e#scrollTo=mC4FRTcQHH4E)
## expermintal changes in dev
 - add Initial Support for Docker
 - add Intial Support for database (broken)
 - experminting with github actions
## How to install ?
### peotry single command
this repo uses poetry as package managent to ease the devlopement so you need to install it first
```bash
pip install poetry
```
and then jsut run the below command to install all needed packages
```bash
poetry install
```
### or  (no-poetry)   
normal `requirements.txt` with command 
```bash
pip install -r requirements.txt
```
## How to run the server ?
this server uses `falcon7B` model and `e5-base-v2` embeddings so it need to download these models  
To run HF Inference Models (models that run on the Hugging Face platform), you need to have at least `~ 2-3 GB`   
of free disk space and RAM to load the models. If you want to download the models locally,
  you should also have enough RAM and hardware capacity, which you can estimate roughly by the model size. 
- first you have to navigate to [HuggingFaceHub](https://huggingface.co/settings/tokens) singup and generate new token   
- create `.env` file in the current directory where the `app.py` live place `huggingfacehub_api_token=YOUR-TOKEN-HERE`
- now go to the command line and type the follwing to run the server
### with poetry
```bash 
poetry run uvicorn main:app 
#or
#poetry run uvicorn main:app --port $PORT 
```
### or (no-peotry)
```bash
uvicorn main:app 
```
- now the server is running to test the api in the command line you can do the next  
first make sure you have curl installed   

**on Linux:**  
```bash 
sudo apt install curl 
```
**on Windows:**  
```bash 
winget install curl
```
## Testing the endpoints
To test the endpoints for this server in your command line, you can use curl commands like this:

- To ask a question over the llm use the `/ask/` endpoint, you can use something like:

```bash
curl http://localhost:8000/ask/?question="your question here"
```

where `your question here` is the question you want to ask.

- To upload files to the `/upload-files/` endpoint, you can use something like:

```bash
curl -L -F "file=@YourFileName" http://localhost:8000/upload_files/
```

where `filename` is the name of the file you want to upload. You can also upload multiple files by adding more `-F` options.

- To perform similarity search on a query using the `/similarity-search/` endpoint, you can use something like:

```bash
curl http://localhost:8000/similarity_search/?query="your query here"
```

where `your query here` is the text you want to search for.

- To ask a question over the data using the `/ask-with-data/` endpoint, you can use something like:

```bash
curl http://localhost:8000/ask_question/?question="your question here"
```

where `your question here` is the question you want to ask.




changelog.md
## version 0.0.2
- using `-` instead of `underscore` for the endpoints
- add `ask/` endpoint for asking the llms directtly without data chatgpt style
- adding changeLog to track changes 

## version 0.0.3
- refactoring the code and big improvment across the codebase.  
- Use dependency injection to create a single instance of the embedding model,
the llm and the prompt template, and inject them into the endpoints that need them.
This would avoid creating them multiple times and make the code more modular and testable.  

- Use FastAPI's background tasks to process the uploaded files and create the embeddings 
asynchronously.This would improve the performance and responsiveness of the server, 
as it would not block the request while doing the heavy computation.  

- Add some error handling and validation to the endpoints,
such as checking if the files are valid text files, if the query or question is not empty
if the vector database is not corrupted, etc. This would make the server more robust and user-friendly.  

- Add some documentation and comments to the code, explaining what each endpoint does, what
parameters it expects and returns, what libraries and models it uses, etc. This would make
the code more readable and maintainable
- add requirements.txt file

## version 0.0.4+dev
- reorganized the code directory
- added docker support - postgress-database + pgadmin  
- dummy database setup not in use yet and not implemented correctly
- add `/api/` route for better consistancy 


api/main.py
import os
from fastapi import FastAPI, File, UploadFile, Depends, BackgroundTasks
from fastapi.responses import JSONResponse
from pathlib import Path
from langchain.document_loaders import UnstructuredFileLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.vectorstores import Chroma
from langchain import PromptTemplate, HuggingFaceHub
from langchain.chains import RetrievalQA

import models
from database import engine, sessionLocal
from dotenv import load_dotenv

load_dotenv()
app = FastAPI()

# Initialize the embedding model and the llm
embedding_model_name = "intfloat/e5-base-v2"
llm_repo_id = "tiiuae/falcon-7b-instruct"

# Define a dependency for creating an embedding model
def get_embedding_model():
    # embedding_model_name = 'intfloat/e5-base-v2'
    embeddings = HuggingFaceEmbeddings(model_name=embedding_model_name)
    return embeddings

# Define a dependency for creating an llm
def get_llm():
    # repo_id = "tiiuae/falcon-7b-instruct"
    llm = HuggingFaceHub(
        repo_id=llm_repo_id,
        huggingfacehub_api_token='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',
        model_kwargs={"temperature":0.6, "max_new_tokens":1024}
    )
    return llm


## setup the database

models.Base.metadata.create_all(bind=engine)
def get_db():
    db = sessionLocal()
    try: yield db
    finally: db.close()

# Define a dependency for creating a prompt template
def get_prompt_template():
    template = """
    You are a helpfull assistant. you gives helpful, summarized, and polite answers to the user's questions given the below context.

    {context}

    question: {question}
    """
    prompt = PromptTemplate(template=template, input_variables=['context', "question"])
    return prompt

# Define a global variable for the vector database
vectordb = None

# Define a function for processing uploaded files and creating embeddings in the background
def process_files(files: list[UploadFile], embeddings: HuggingFaceEmbeddings):
    global vectordb # use the global variable
    # Create loaders for each file
    loaders = [UnstructuredFileLoader(file.file) for file in files]
    # Load the documents from each loader
    documents = [loader.load()[0] for loader in loaders]
    # Split the documents into text parts using RecursiveCharacterTextSplitter
    docs = RecursiveCharacterTextSplitter(
        chunk_size=750,
        chunk_overlap=100,
        separators=[",", "\n", "\n\n"]
    ).split_documents(documents)
    # Create a chroma database from the documents and embeddings
    vectordb = Chroma.from_documents(
        documents=docs,
        embedding=embeddings,
        persist_directory='db'
    )





# Define a global variable for the vector database
vectordb = None


# Define an endpoint for uploading text files and creating embeddings
@app.post("/api/upload_files/")
async def upload_files(
    files: list[UploadFile] = File(...),
    embeddings: HuggingFaceEmbeddings = Depends(get_embedding_model),
    background_tasks: BackgroundTasks = BackgroundTasks()
):
    # Validate that all files are text files by checking their content type
    for file in files:
        if not file.content_type.startswith("text/"):
            return JSONResponse(status_code=400, content={"message": f"Invalid file type: {file.content_type}"})
    
    # Add a background task to process the files and create embeddings asynchronously
    background_tasks.add_task(process_files, files, embeddings)

    # Return a success message with the number of documents processed
    return {"message": f"Processed {len(files)} documents"}


# Define an endpoint for performing similarity search on a query
@app.get("/api/similarity_search/")
async def similarity_search(
    query: str,
    embeddings: HuggingFaceEmbeddings = Depends(get_embedding_model)
):
    global vectordb # use the global variable
    # Check if the vector database is initialized
    if vectordb is None:
        return JSONResponse(status_code=400, content={"message": "No documents uploaded yet"})
    # Check if the query is not empty
    if not query:
        return JSONResponse(status_code=400, content={"message": "Query cannot be empty"})
    # Perform similarity search on the query using the vector database
    docs = vectordb.similarity_search(query)
    # Return a list of matching documents with their metadata and page content
    return {"results": [{"metadata": doc.metadata,
                        "page_content": doc.page_content} for doc in docs]}



# Define an endpoint for asking questions over the data using RetrievalQA
@app.get("/api/ask-with-data/")
async def ask_question_with_data(
    question: str,
    llm: HuggingFaceHub = Depends(get_llm),
    prompt: PromptTemplate = Depends(get_prompt_template)
    ):
    global vectordb  # use the global variable
# Check if the vector database is initialized
    if vectordb is None:
        return JSONResponse(status_code=400, content={"message": "No documents uploaded yet"})
    # Check if the question is not empty
    if not question:
        return JSONResponse(status_code=400, content={"message": "Question cannot be empty"})
    # Create a RetrievalQA object using the llm, vector database and prompt template
    qa = RetrievalQA.from_chain_type(
        llm=llm,
        chain_type="stuff",
        retriever=vectordb.as_retriever(),
        chain_type_kwargs={"prompt": prompt},
        return_source_documents=True,
    )
    # Ask the question using the RetrievalQA object and get the results
    results = qa(question)
    # Return the answer and the source documents used to generate it
    return {
        "answer": results['result'],
        "source_documents": results['source_documents']
    }


# Define an endpoint for asking questions over the data using RetrievalQA
@app.get("/api/ask/")
async def ask_question(
    question: str,
    llm: HuggingFaceHub = Depends(get_llm),
    ):

    template = """
You are a helpfull assistant. you gives helpful, summarized, and polite answers to the user's questions.
question: {question}
""".format(question=question)
    
    results = llm(template)
    return {
        "answer": results
    }

app.py
import os
from fastapi import FastAPI, File, UploadFile, Depends, BackgroundTasks
from fastapi.responses import JSONResponse
from pathlib import Path
from langchain.document_loaders import UnstructuredFileLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.vectorstores import Chroma
from langchain import PromptTemplate, HuggingFaceHub
from langchain.chains import RetrievalQA
from dotenv import load_dotenv

load_dotenv()
app = FastAPI()

# Initialize the embedding model and the llm
embedding_model_name = "intfloat/e5-base-v2"
llm_repo_id = "tiiuae/falcon-7b-instruct"

# Define a dependency for creating an embedding model
def get_embedding_model():
    # embedding_model_name = 'intfloat/e5-base-v2'
    embeddings = HuggingFaceEmbeddings(model_name=embedding_model_name)
    return embeddings

# Define a dependency for creating an llm
def get_llm():
    # repo_id = "tiiuae/falcon-7b-instruct"
    llm = HuggingFaceHub(
        repo_id=llm_repo_id,
        huggingfacehub_api_token='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',
        model_kwargs={"temperature":0.6, "max_new_tokens":1024}
    )
    return llm

# Define a dependency for creating a prompt template
def get_prompt_template():
    template = """
    You are a helpfull assistant. you gives helpful, summarized, and polite answers to the user's questions given the below context.

    {context}

    question: {question}
    """
    prompt = PromptTemplate(template=template, input_variables=['context', "question"])
    return prompt

# Define a global variable for the vector database
vectordb = None

# Define a function for processing uploaded files and creating embeddings in the background
def process_files(files: list[UploadFile], embeddings: HuggingFaceEmbeddings):
    global vectordb # use the global variable
    # Create loaders for each file
    loaders = [UnstructuredFileLoader(file.file) for file in files]
    # Load the documents from each loader
    documents = [loader.load()[0] for loader in loaders]
    # Split the documents into text parts using RecursiveCharacterTextSplitter
    docs = RecursiveCharacterTextSplitter(
        chunk_size=750,
        chunk_overlap=100,
        separators=[",", "\n", "\n\n"]
    ).split_documents(documents)
    # Create a chroma database from the documents and embeddings
    vectordb = Chroma.from_documents(
        documents=docs,
        embedding=embeddings,
        persist_directory='db'
    )





# Define a global variable for the vector database
vectordb = None


# Define an endpoint for uploading text files and creating embeddings
@app.post("/upload_files/")
async def upload_files(
    files: list[UploadFile] = File(...),
    embeddings: HuggingFaceEmbeddings = Depends(get_embedding_model),
    background_tasks: BackgroundTasks = BackgroundTasks()
):
    # Validate that all files are text files by checking their content type
    for file in files:
        if not file.content_type.startswith("text/"):
            return JSONResponse(status_code=400, content={"message": f"Invalid file type: {file.content_type}"})
    
    # Add a background task to process the files and create embeddings asynchronously
    background_tasks.add_task(process_files, files, embeddings)

    # Return a success message with the number of documents processed
    return {"message": f"Processed {len(files)} documents"}


# Define an endpoint for performing similarity search on a query
@app.get("/similarity_search/")
async def similarity_search(
    query: str,
    embeddings: HuggingFaceEmbeddings = Depends(get_embedding_model)
):
    global vectordb # use the global variable
    # Check if the vector database is initialized
    if vectordb is None:
        return JSONResponse(status_code=400, content={"message": "No documents uploaded yet"})
    # Check if the query is not empty
    if not query:
        return JSONResponse(status_code=400, content={"message": "Query cannot be empty"})
    # Perform similarity search on the query using the vector database
    docs = vectordb.similarity_search(query)
    # Return a list of matching documents with their metadata and page content
    return {"results": [{"metadata": doc.metadata,
                        "page_content": doc.page_content} for doc in docs]}



# Define an endpoint for asking questions over the data using RetrievalQA
@app.get("/ask-with-data/")
async def ask_question_with_data(
    question: str,
    llm: HuggingFaceHub = Depends(get_llm),
    prompt: PromptTemplate = Depends(get_prompt_template)
    ):
    global vectordb  # use the global variable
# Check if the vector database is initialized
    if vectordb is None:
        return JSONResponse(status_code=400, content={"message": "No documents uploaded yet"})
    # Check if the question is not empty
    if not question:
        return JSONResponse(status_code=400, content={"message": "Question cannot be empty"})
    # Create a RetrievalQA object using the llm, vector database and prompt template
    qa = RetrievalQA.from_chain_type(
        llm=llm,
        chain_type="stuff",
        retriever=vectordb.as_retriever(),
        chain_type_kwargs={"prompt": prompt},
        return_source_documents=True,
    )
    # Ask the question using the RetrievalQA object and get the results
    results = qa(question)
    # Return the answer and the source documents used to generate it
    return {
        "answer": results['result'],
        "source_documents": results['source_documents']
    }


# Define an endpoint for asking questions over the data using RetrievalQA
@app.get("/ask/")
async def ask_question(
    question: str,
    llm: HuggingFaceHub = Depends(get_llm),
    ):

    template = """
You are a helpfull assistant. you gives helpful, summarized, and polite answers to the user's questions.
question: {question}
""".format(question=question)
    
    results = llm(template)
    return {
        "answer": results
    }